File: netDxf/IO/DxfWriter.Sun.cs
using System.Linq;
using netDxf.Collections;
using netDxf.Objects;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private void PrepareSunClass(DxfClassCollection definitions)
        { this.PrepareStoredEnvelopeClass(definitions, "SUN", "AcDbSun", "SCENEOE", 1153, this.doc.Objects.Items.Any(item => item is DxfSun)); }
        private bool WriteSunPayload(DxfDatabaseObject item)
        {
            if (!(item is DxfSun sun)) return false;
            this.chunk.Write(100, "AcDbSun");
            this.chunk.Write(90, sun.StoredVersion);
            this.chunk.Write(290, sun.Enabled);
            this.chunk.Write(63, sun.ColorIndex);
            if (sun.TrueColor.HasValue) this.chunk.Write(421, sun.TrueColor.Value);
            this.chunk.Write(40, sun.Intensity);
            this.chunk.Write(291, sun.ShadowsEnabled);
            this.chunk.Write(91, sun.JulianDay);
            this.chunk.Write(92, sun.StoredTime);
            this.chunk.Write(292, sun.DaylightSavingTime);
            this.chunk.Write(70, (short)sun.ShadowType);
            this.chunk.Write(71, sun.ShadowMapSize);
            this.chunk.Write(280, (short)sun.ShadowSoftness);
            return true;
        }
        private void WriteSunReference(DxfObject owner)
        {
            if (!SunReferences.IsPresent(owner)) return;
            SunReferences.CheckProfile(owner, this.doc.DrawingVariables.AcadVer);
            this.chunk.Write(361, SunReferences.Get(owner)?.Handle ?? "0");
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.Sun.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfSun } from '../Objects/DxfSun.js';
import { SunReferences } from '../Objects/SunReferences.js';
import { PrepareStoredEnvelopeClass } from './DxfWriter.LayerFilterPointer.js';
export function PrepareSunClass(document,definitions) {
  PrepareStoredEnvelopeClass(document,definitions,'SUN','AcDbSun','SCENEOE',1153,Array.from(document.Objects.Items).some(item=>item instanceof DxfSun));
}
export function WriteSunPayload(chunk,version,item) {
  if(!(item instanceof DxfSun))return false;
  chunk.Write(100,'AcDbSun');chunk.Write(90,item.StoredVersion);chunk.Write(290,item.Enabled);chunk.Write(63,item.ColorIndex);
  if(item.TrueColor!==null)chunk.Write(421,item.TrueColor);
  chunk.Write(40,item.Intensity);chunk.Write(291,item.ShadowsEnabled);chunk.Write(91,item.JulianDay);chunk.Write(92,item.StoredTime);chunk.Write(292,item.DaylightSavingTime);
  chunk.Write(70,(item.ShadowType<<16)>>16);chunk.Write(71,item.ShadowMapSize);chunk.Write(280,item.ShadowSoftness);return true;
}
export function WriteSunReference(chunk,version,owner) {
  if(!SunReferences.IsPresent(owner))return;
  SunReferences.CheckProfile(owner,version);chunk.Write(361,SunReferences.Get(owner)?.Handle??'0');
}
